proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,259 @@
+computed from the rewards in the block as follows.
+
+```
+reserve_hash = hash([(stake_account_key, reward_amount)])
+```
+
+At the end of reward computation phase, M `EpochRewardsReserve` sysvar
+accounts are updated, where `M` corresponds to how many blocks will be in
+the reward distribution phase, which is described in the next section.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks.
+
+Based on current validator client's performance, 64 additional accounts can be
+stored per tick time without much performance impact for the overall block
+time. Given that each block time is 64 ticks time long, a total of 4,096
+accounts can be distributed per block. The total number of blocks needed tto
+distributed rewards is
+
+```
+M = ((4096 - 1)+num_stake_accounts)/4096
+```
+
+During the reward distribution phase, for the block at `block_height`, the
+rewards will be distributed from the `EpochRewardReserve` sysvar account with
+the address of "hash(`SysvarRewardReserve11111111111111111111111111111111`,
+block_height). As the stake_accounts receive their rewards from the reserve
+account, the balance of the EpochRewardReserve will decrease. Once all rewards
+have been distributed, the balance of the EpochRewardReserve account should be
+reduced to 0.
+
+Before each reward distribution, the `EpochRewardReserve` account's balance
+is checked to make sure there is enough balance to distribute the reward. After a
+reward is distributed, the balance in `EpochRewardReserve` account is decreased
+by the amount of the reward that was distributed. The hash of reward_account
+pubkey and reward amount is accumulated when rewards are distributed. After all
+the rewards are distributed from the reserve account for the block, the hash
+computed from all the rewards distribution is verified against the hash stored
+in the `EpochRewardReserve` account to make sure they match.
+
+During the reward distribution phase, there is a "root accumulator hash" that
+starts empty and is gradually filled by adding each hash of a
+`EpochRewardReserve`. The distribution process continues until the "root
+accumulator hash" matches the "root hash" stored in `EpochRewardHistory`. At this
+point, the reward distribution for the current epoch is complete and we can
+safely assume that there will be no further distributions for the remainder of
+the epoch.
+
+To ensure that each block distributes a subset of the rewards in a deterministic
+manner for the current epoch, while also randomizing the distribution across
+different epochs, the partitioning of all rewards will be done in a specific
+way. This will involve hashing the account's public key with the epoch number,
+sorting the resulting values, and dividing them into `M` bins. This process will
+ensure that the reward distribution is deterministic for the current epoch,
+while also introducing a degree of randomness between epochs.
```

Comment:
Hmm, isn't this a fundamental part of the rewards calculation process? Why is it not included in that section? Feels like an afterthought, way down here.
Where is the partitioned list of stake account keys stored?